flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java

```diff
@@ -117,80 +118,61 @@ public void testBufferUsageMetrics() throws IOException, InterruptedException {
 
-			List<Buffer> exclusiveBuffers = new ArrayList<>();
 
 			// test for exclusive buffer used
-			exclusiveBuffers.add(remoteInputChannel1.requestBuffer());
-			assertEquals(1, exclusiveBuffersUsageGauge.calculateUsedBuffers(inputGate1));
-
-			for (int i = 0; i < buffersPerChannel; i++){
-				Buffer buffer = remoteInputChannel1.requestBuffer();
-				if (buffer != null) {
-					exclusiveBuffers.add(buffer);
-				} else {
-					break;
-				}
-			}
-			assertEquals(buffersPerChannel, exclusiveBuffers.size());
+			drainBuffer(buffersPerChannel, remoteInputChannel1);
+
 			assertEquals(buffersPerChannel, exclusiveBuffersUsageGauge.calculateUsedBuffers(inputGate1));
 			assertEquals(1 / 3.0, exclusiveBuffersUsageGauge.getValue(), 0.0001);
 			assertEquals(buffersPerChannel / (extraNetworkBuffersPerGate * 2 + buffersPerChannel * 3.0), inputBufferPoolUsageGauge.getValue(), 0.0001);
 
-			for (int i = 0; i < buffersPerChannel; i++){
-				Buffer buffer = remoteInputChannel2.requestBuffer();
-				if (buffer != null) {
-					exclusiveBuffers.add(buffer);
-				} else {
-					break;
-				}
-			}
-			assertEquals(buffersPerChannel * 2, exclusiveBuffers.size());
+			drainBuffer(buffersPerChannel, remoteInputChannel2);
+
 			assertEquals(buffersPerChannel * 2, exclusiveBuffersUsageGauge.calculateUsedBuffers(inputGate1));
 			assertEquals(2 / 3.0, exclusiveBuffersUsageGauge.getValue(), 0.0001);
 			assertEquals(buffersPerChannel * 2 / (extraNetworkBuffersPerGate * 2 + buffersPerChannel * 3.0), inputBufferPoolUsageGauge.getValue(), 0.0001);
 
 			// test for floating buffer used
-			List<Buffer> floatingBuffers = new ArrayList<>();
-
 			remoteInputChannel1.requestSubpartition(0);
 			remoteInputChannel1.onSenderBacklog(3);
 			assertEquals(buffersPerChannel + 3, remoteInputChannel1.unsynchronizedGetFloatingBuffersAvailable());
 
-			for (int i = 0; i < buffersPerChannel + 3; i++) {
-				Buffer buffer = remoteInputChannel1.requestBuffer();
-				if (buffer != null) {
-					floatingBuffers.add(buffer);
-				} else {
-					break;
-				}
-			}
+			drainBuffer(buffersPerChannel + 3, remoteInputChannel1);
+
 			assertEquals(0, remoteInputChannel1.unsynchronizedGetFloatingBuffersAvailable());
 			assertEquals(buffersPerChannel * 2, exclusiveBuffersUsageGauge.calculateUsedBuffers(inputGate1));
 			assertEquals(2 / 3.0, exclusiveBuffersUsageGauge.getValue(), 0.0001);
 			assertEquals((buffersPerChannel * 2 + buffersPerChannel + 3) / (extraNetworkBuffersPerGate * 2 + buffersPerChannel * 3.0), inputBufferPoolUsageGauge.getValue(), 0.0001);
 
 			List<MemorySegment> segments = new ArrayList<>();
-			for (Buffer buffer : exclusiveBuffers) {
-				segments.add(buffer.getMemorySegment());
-			}
+
 			network.getNetworkBufferPool().recycleMemorySegments(segments);
 		} finally {
 			inputGate1.close();
 			inputGate2.close();
-			network.shutdown();
+			closeableRegistry.close();
+			network.close();
```

Comment:
nit: you could add `network.close()` and `inputGateN.close()` to the `closeableRegistry`

Reply:
I think this is not so good, because the closeableRegistry do not close respect to the register order which may lead to network closed before network buffers has all be recycled.